### ISSUE TITLE
Laggy First Keyboard Open (Physical iOS Device)

### ISSUE THREAD

**Issue by willsmanley** [flutter/flutter#160725]
### Steps to reproduce

checkout https://github.com/willsmanley/flutter-textfield-keyboard-ios-bug

run on a physical ios device (with either impeller or skia)

press the textfield to focus it and trigger a keyboard open. the first time during app run takes 2 second lag. every subsequent time is immediate.

### Expected results

should open immediately like native swift apps

### Actual results

opens keyboard after 1-2 second delay

### Code sample

https://github.com/willsmanley/flutter-textfield-keyboard-ios-bug

### Logs

n/a

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.27.0, on macOS 15.1 24B2082 darwin-arm64, locale en-US)
[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/to/macos-android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.

[✓] Xcode - develop for iOS and macOS (Xcode 16.2)
[✓] Chrome - develop for the web
[!] Android Studio (not installed)
[✓] Connected device (6 available)
[✓] Network resources

! Doctor found issues in 2 categories.
```

</details>

**Comment by danagbemava-nc** [comment]
Hi @willsmanley, do you get the same issue in release mode or is this only in debug mode?

**Comment by willsmanley** [comment]
Also in profile/release mode

**Comment by danagbemava-nc** [comment]
Hi @willsmanley, does this video capture your issue?

https://github.com/user-attachments/assets/0c15c4a6-2085-4884-862a-32ef86d8e464

**Comment by willsmanley** [comment]
I think so! Although I can't tell when the text field was tapped. 

Another good way to see it very clearly is if you navigate to a new screen with an autofocus text field. If it's the first time opening, it will jank through the animation. Subsequent navigations paired with textfield autofocus will be smooth as usual.

**Comment by danagbemava-nc** [comment]
Hi @willsmanley, can you check on the master channel to see if you still experience this there? From my testing, it appears that this has either been fixed or at least greatly improved on the master channel. I added timeDilation so the effect can be better seen


https://github.com/user-attachments/assets/430c037f-2e35-4d28-97e6-6d2bbba9e6d9



<details>
<summary>sample tested</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

void main() {
  timeDilation = 10;
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Builder(builder: (context) {
        return Scaffold(
          body: Center(
            child: TextButton(
              onPressed: () {
                Navigator.of(context).push(
                  MaterialPageRoute(
                    builder: (context) => Scaffold(
                      body: Center(
                        child: TextField(
                          autofocus: true,
                        ),
                      ),
                    ),
                  ),
                );
              },
              child: Text('Go to Next'),
            ),
          ),
        );
      }),
    );
  }
}
```
</details>

**Comment by willsmanley** [comment]
I tried switching to master channel in my production app, which still seems to have the issue. 

But I will try again on the minimal repro app in release mode tomorrow.

**Comment by willsmanley** [comment]
ok so i had a chance to check it. fortunately, it seems to only occur on debug and profile mode. release mode looks fine.

still i think it would be a good thing to fix but at least it is not affecting production. 

im seeing the same behavior between master and 3.24/3.27 branches.

**Comment by danagbemava-nc** [comment]
> im seeing the same behavior between master and 3.24/3.27 branches.

This is in profile & debug mode, correct?

**Comment by willsmanley** [comment]
Yes but not in release mode

**Comment by danagbemava-nc** [comment]
Thanks for confirming. 
Reproducible using the sample in https://github.com/flutter/flutter/issues/160725#issuecomment-2568771135

The behavior is better on the master channel in release mode but aside that, the behavior is the same on all modes (release, debug, profile) on stable & (profile & debug) on master. 

Labeling for further investigation

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 3.27.1, on macOS 15.1.1 24B91 darwin-arm64, locale en-US)
    • Flutter version 3.27.1 on channel stable at /Users/deanli/dev/stable
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 17025dd882 (3 weeks ago), 2024-12-17 03:23:09 +0900
    • Engine revision cb4b5fff73
    • Dart version 3.6.0
    • DevTools version 2.40.2
```

```
[!] Flutter (Channel master, 3.28.0-2.0.pre.38644, on macOS 15.1.1 24B91 darwin-arm64, locale en-US)
    • Flutter version 3.28.0-2.0.pre.38644 on channel master at /Users/deanli/dev/master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 5f2bf18183 (6 hours ago), 2025-01-07 17:48:30 -0800
    • Engine revision 5f2bf18183
    • Dart version 3.7.0 (build 3.7.0-267.0.dev)
    • DevTools version 2.41.0

```

</details>

**Comment by justinmc** [comment]
Well at least this works better in release mode. Nothing recent off the top of my head would affect this, but I'm not sure.

**Comment by hellohuanlin** [comment]
What's your device? Personally I have noticed keyboard being slow even for native apps on older devices almost every iOS update.

**Comment by willsmanley** [comment]
iPhone 15 pro max, but it happens on simulator too

**Comment by Amber916Young** [comment]
same here, iPhone Se2

**Comment by RCVZ** [comment]
This issue is still present in version 3.29. Please investigate.

**Comment by chrisfranko** [comment]
Im noticing this issue on my iphone 16 pro. latest version of flutter in debug mode and im also getting this warning.

```
Error: this application, or a library it uses, has passed an invalid numeric value (NaN, or not-a-number) to CoreGraphics API and this value is being ignored. Please fix this problem.
If you want to see the backtrace, please set CG_NUMERICS_SHOW_BACKTRACE environmental variable.
```

**Comment by RCVZ** [comment]
@danagbemava-nc could you set this to p1 since it is still present in version 3.29

**Comment by Rupsero** [comment]
Yep, issue does still exist. Has anyone found a solution for this?

**Comment by john7964** [comment]
Check `FocusScope.of(context).hasFocus == true`
I faced the same problem and I found that default FocusScope has focused initially.
And then I wrapped the TextField with a FocusScope fixed the problem.

**Comment by willsmanley-archive** [comment]
@john7964 I believe you are just deferring the lag issue from the first time the page renders (due to keyboard focus default) to the first time the keyboard is focused after that. I have found that, on master channel, it is still laggy on first keyboard open.

**Comment by kkoskenvirta** [comment]
This would be amazing if it would get fixed. It is a shame that it has been happening for a long time now.

**Comment by danagbemava-nc** [comment]
For anyone else encountering this issue, please add a thumbs up to the issue instead. Kindly read [https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md#comments](https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md#comments) and https://github.com/flutter/flutter/blob/master/docs/contributing/issue_hygiene/README.md#thumbs-up-reactions

Thank you

**Comment by kkoskenvirta** [comment]
I think I found a possible root cause for the issue. In the 2023 WWDC, Apple reworked the keyboard architecture. At the end of the explanation video, they mention that the new architecture might cause issues if the main thread is busy during the keyboard opening process — [see this point in the video.](https://youtu.be/49VhMsZoa6E?t=805)

This could explain why people are noticing the issue more often when using bottom sheets with autofocus set on a TextField. In my case, I observed that when a bottom sheet contained an autofocus TextField, the entire root Scaffold would re-render as the keyboard opened — even when resizeToAvoidBottomInset was set to false.

After completely removing the Scaffold, I now have a smooth experience. By also minimizing main thread activity during keyboard open, I’ve been able to achieve a noticeably smoother transition.

**Comment by christianarduino** [comment]
Any updates?
